app/assets/javascripts/discourse/tests/helpers/qunit-helpers.js

```diff
@@ -610,14 +605,28 @@ export async function paste(element, text, otherClipboardData = {}) {
   return e;
 }
 
-export async function emulateAutocomplete(inputSelector, text) {
-  await triggerKeyEvent(inputSelector, "keydown", "Backspace");
-  await fillIn(inputSelector, `${text} `);
-  await triggerKeyEvent(inputSelector, "keyup", "Backspace");
+export async function simulateKey(element, key) {
+  const keyCode = key.charCodeAt(0);
+  const event = { key, keyCode, which: keyCode };
+
+  element.dispatchEvent(new KeyboardEvent("keydown", event));
 
-  await triggerKeyEvent(inputSelector, "keydown", "Backspace");
-  await fillIn(inputSelector, text);
-  await triggerKeyEvent(inputSelector, "keyup", "Backspace");
+  const pos = element.selectionStart;
+  const value = element.value;
+
+  if (key === "\b") {
+    element.value = value.slice(0, pos - 1) + value.slice(pos);
+    element.selectionStart = pos - 1;
+    element.selectionEnd = pos - 1;
+  } else if (key !== "\r") {
```

Comment:
To better simulate the behaviour that is happening in the browser, we don't "add" the <kbd>Enter</kbd> character in the `element`.

It will be caught by the `autocomplete` library though.

Reply:
This helper function can be simplified into:

```js
  if (key === "\b") {
    await triggerKeyEvent(element, "keydown", "Backspace");

    const pos = element.selectionStart;
    element.value = element.value.slice(0, pos - 1) + element.value.slice(pos);
    element.selectionStart = pos - 1;
    element.selectionEnd = pos - 1;

    await triggerKeyEvent(element, "keyup", "Backspace");
  } else if (key !== "\r") {
    await typeIn(element, key);
  } else {
    await triggerKeyEvent(element, "keydown", "Enter");
    await triggerKeyEvent(element, "keyup", "Enter");
  }
```